### DISCUSSION TITLE
How to fix reportAssignmentType when using a function that return a tuple?

### DISCUSSION THREAD

**Discussion by alanlivio** [microsoft/pyright#7257] [Q&A] [answered]
I received a reportAssignmentType  in the code bellow. Even the explicit cast `: NDArray` does not avoid it.
```
from numpy.typing import NDArray
from typing import Union

def request() -> Union[tuple[NDArray,float],NDArray]:
    ...

ret: NDArray = request()
```

```
Expression of type "tuple[NDArray[Unknown], float] | NDArray[Unknown]" cannot be assigned to declared type "NDArray[Unknown]"
  Type "tuple[NDArray[Unknown], float] | NDArray[Unknown]" cannot be assigned to type "NDArray[Unknown]"
    "tuple[NDArray[Unknown], float]" is incompatible with "NDArray[Unknown]"Pylance[reportAssignmentType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportAssignmentType)
```

```
$ python --version
Python 3.9.18
$ pyright --version
pyright 1.1.350
```

I appreciate any suggestion.

**Comment by erictraut** [comment, marked as answer]
Pyright's behavior here is correct. You're attempting to assign a value of type `tuple[NDArray, float] | NDArray` to a value that is defined to accept `NDArray` (or subtypes thereof). The value you're assigning is a wider type than is allowed by `ref`, so pyright is correct to generate a type violation error here.

I don't know enough about what you're trying to do here to suggest a workaround. If the `request` function can return either a `tuple` or an `NDArray` but you're assuming it always returns an `NDArray`, you can add an assert to validate this assumption.

```python
from numpy.typing import NDArray

def request() -> tuple[NDArray, float] | NDArray: ...

ret = request()
assert not isinstance(ret, tuple)
reveal_type(ret) # ndarray
```

**Comment by alanlivio** [comment]
Thanks @erictraut. Indeed the assert works and ensure the type.
